File: binary_tree/valid_bst.rb
# 98
# Inorder traversal approach
def is_valid_bst(root)
  is_valid_bst_helper(root, TreeNode.new(-Float::INFINITY))
end

def is_valid_bst_helper(root, previous)
  # recursive base case
  return true if root.nil?

  left, right = root.left, root.right

  # perform operation on left child recursively
  is_valid_bst_helper(left, root)

  # perform operation on current node
  if previous.val >= root.val
      return false
  end

  # perform operation on right child
  is_valid_bst_helper(right, previous)
end

File: binary_tree/valid_bst.js
// https://leetcode.com/problems/validate-binary-search-tree/description/
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
  if (!root) return true;
  const minHolder = [Number.NEGATIVE_INFINITY];
  return inOrderTraversal(root, minHolder);
};

function inOrderTraversal(root, minHolder) {
  let left = true;
  let right = true;
  let current;

  if (root.left) left = inOrderTraversal(root.left, minHolder);

  if (root.val <= minHolder[0]) {
    current = false;
  } else {
    minHolder[0] = root.val;
    current = true;
  }

  if (root.right) right = inOrderTraversal(root.right, minHolder);

  return left && current && right;
}
